go/cmd/dolt/commands/branch.go

```diff
@@ -103,91 +109,166 @@ func (cmd BranchCmd) Exec(ctx context.Context, commandStr string, args []string,
 	help, usage := cli.HelpAndUsagePrinters(cli.CommandDocsForCommandString(commandStr, branchDocs, ap))
 	apr := cli.ParseArgsOrDie(ap, args, help)
 
+	queryEngine, sqlCtx, closeFunc, err := cliCtx.QueryEngine(ctx)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.BuildDError("error: failed to create query engine").AddCause(err).Build(), nil)
+	}
+
+	if closeFunc != nil {
+		defer closeFunc()
+	}
+
+	if len(apr.ContainsMany(cli.MoveFlag, cli.CopyFlag, cli.DeleteFlag, cli.DeleteForceFlag, cli.ListFlag, showCurrentFlag)) > 1 {
+		cli.PrintErrln("Must specify exactly one of --move/-m, --copy/-c, --delete/-d, -D, --show-current, or --list.")
+		return 1
+	}
+
+	for _, arg := range apr.Args {
+		if !doltdb.IsValidUserBranchName(arg) {
+			cli.PrintErrf("%s is an invalid branch name", arg)
+		}
+	}
+
 	switch {
 	case apr.Contains(cli.MoveFlag):
-		return moveBranch(ctx, dEnv, apr, usage)
+		return moveBranch(sqlCtx, queryEngine, apr, args, usage)
 	case apr.Contains(cli.CopyFlag):
-		return copyBranch(ctx, dEnv, apr, usage)
+		return copyBranch(sqlCtx, queryEngine, apr, args, usage)
 	case apr.Contains(cli.DeleteFlag):
-		return deleteBranches(ctx, dEnv, apr, usage, apr.Contains(cli.ForceFlag))
+		return deleteBranches(sqlCtx, queryEngine, apr, args, usage)
 	case apr.Contains(cli.DeleteForceFlag):
-		return deleteBranches(ctx, dEnv, apr, usage, true)
+		return deleteBranches(sqlCtx, queryEngine, apr, args, usage)
 	case apr.Contains(cli.ListFlag):
-		return printBranches(ctx, dEnv, apr, usage)
+		return printBranches(sqlCtx, queryEngine, apr, usage)
 	case apr.Contains(showCurrentFlag):
-		return printCurrentBranch(dEnv)
+		return printCurrentBranch(sqlCtx, queryEngine)
 	case apr.Contains(datasetsFlag):
 		return printAllDatasets(ctx, dEnv)
 	case apr.NArg() > 0:
-		return createBranch(ctx, dEnv, apr, usage)
+		return createBranch(sqlCtx, queryEngine, apr, args, usage)
 	default:
-		return printBranches(ctx, dEnv, apr, usage)
+		return printBranches(sqlCtx, queryEngine, apr, usage)
+	}
+}
+
+type branchMeta struct {
+	name   string
+	hash   string
+	remote bool
+}
+
+func getBranches(sqlCtx *sql.Context, queryEngine cli.Queryist, remote bool) ([]branchMeta, error) {
```

Comment:
Are there other commands which will need this? Possibly put it in utils.go, but ok to punt until we need it.

Reply:
Maybe, but `dolt branch` is the canonical way to list branches, so until we need to, it makes sense to put it here.